Asynchronous polling engine & benchmarking tool (#6030)
OpenNMS/opennms: docs/modules/operation/pages/deep-dive/service-assurance/configuration.adoc

* Node metadata
* Interface metadata
* Service metadata

[[ga-pollerd-configuration-async]]
== Asynchronous engine

By default, the number of polls active at any given time is limited by the number of threads that are allocated to the daemon.

> Each thread is responsible for one poll, and will block until the poll is completed.
> This is true, even when the polling is done by a Minion at a remote location.

Reviewer: ```suggestion
Each thread is responsible for one poll, and will block until the poll is completed; this is true even when the polling is done by a Minion at a remote location.
```

> In this case, the thread will block for the duration of the RPC.

Reviewer: Here, are we using RPC as an industry-standard term (i.e. Remote Procedure Call)?

```suggestion
In this case, the thread is blocked for the duration of the RPC.
```

> In scenarios where there are hundreds of different locations, and hundreds of nodes at each of these, it can be desirable to reduce the number of threads required on the core.

Reviewer: ```suggestion
If you have hundreds of locations and hundreds of nodes at each location, you may want to reduce the number of threads required on your {page-component-title} instance.
```

> Each thread allocates a fixed amount of system memory, and too many threads active at a given time can extraneous load on the system due to context switching.

Reviewer: ```suggestion
Each thread allocates a fixed amount of system memory.
Having too many threads active at a given time can cause extensive load on the system.
```

> For these reasons, we have developed a new asynchronous mode for the poller.
> When enabled, threads are released while the polls are active, allowing Minions to process more polls than there are threads.

Reviewer: ```suggestion
When enabled, the asynchronous polling mode mitigates this load.
It allows threads to be released while polls are active, enabling Minions to process more polls than the number of threads available.
```

> To enable the feature, set the following attributes in the top level element of `etc/poller-configuration.xml`:

Reviewer: ```suggestion
To enable the feature, set the following attributes in the top-level element of `etc/poller-configuration.xml`:

```

[source, xml]
----
<poller-configuration threads="30"
                      asyncPollingEngineEnabled="true" <1>
                      maxConcurrentAsyncPolls="200" <2>

> ----
> 
> <1> Enable the asynchronous engine.

Reviewer: ```suggestion
----
<1> Enables the asynchronous polling engine.
```

> <2> Used to limit the number of polls active, since each of these still consume some memory resources.

Reviewer 2: ```suggestion
<2> Used to limit the number of polls active, since each of these still consumes some memory resources.
```

> In this mode, threads are only used to trigger the execution of the poll, so the number can be typically reduced.

Reviewer: ```suggestion
In this mode, threads are used only to trigger a poll to run; this means that the number of threads can typically be reduced.
```

> The actual number of polls active, or in-flight, can be observed via the `NumPollsInFlight` metrics associated with the daemon's MBeans.

Reviewer: ```suggestion
You can see the actual number of active (in-flight) polls by reviewing the `NumPollsInFlight` metrics associated with pollerd's MBeans.
```
